rclpy/src/rclpy/action_server.cpp

```diff
@@ -0,0 +1,418 @@
+  /* taken_msg is always destroyed in this function */ \
+  auto taken_msg = create_from_py(pymsg_type); \
+  rmw_request_id_t header; \
+  rcl_ret_t ret = \
+    rcl_action_take_ ## Type ## _request(rcl_action_server_.get(), &header, taken_msg.get()); \
+  /* Create the tuple to return */ \
+  if (ret == RCL_RET_ACTION_CLIENT_TAKE_FAILED || ret == RCL_RET_ACTION_SERVER_TAKE_FAILED) { \
+    return py::make_tuple(py::none(), py::none()); \
+  } else if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to take " #Type); \
+  } \
+  return py::make_tuple(header, convert_to_py(taken_msg.get(), pymsg_type)); \
+
+py::tuple
+ActionServer::take_goal_request(py::object pymsg_type)
+{
+  TAKE_SERVICE_REQUEST(goal)
+}
+
+py::tuple
+ActionServer::take_result_request(py::object pymsg_type)
+{
+  TAKE_SERVICE_REQUEST(result)
+}
+
+#define SEND_SERVICE_RESPONSE(Type) \
+  auto ros_response = convert_from_py(pyresponse); \
+  rcl_ret_t ret = rcl_action_send_ ## Type ## _response( \
+    rcl_action_server_.get(), header, ros_response.get()); \
+  if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to send " #Type " response"); \
+  }
+
+void
+ActionServer::send_goal_response(
+  rmw_request_id_t * header, py::object pyresponse)
+{
+  SEND_SERVICE_RESPONSE(goal)
+}
+
+void
+ActionServer::send_result_response(
+  rmw_request_id_t * header, py::object pyresponse)
+{
+  SEND_SERVICE_RESPONSE(result)
+}
+
+py::tuple
+ActionServer::take_cancel_request(py::object pymsg_type)
+{
+  TAKE_SERVICE_REQUEST(cancel)
+}
+
+void
+ActionServer::send_cancel_response(
+  rmw_request_id_t * header, py::object pyresponse)
+{
+  SEND_SERVICE_RESPONSE(cancel)
+}
+
+void
+ActionServer::publish_feedback(py::object pymsg)
+{
+  auto ros_message = convert_from_py(pymsg);
+  rcl_ret_t ret = rcl_action_publish_feedback(rcl_action_server_.get(), ros_message.get());
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to publish feedback with an action server");
+  }
+}
+
+void
+ActionServer::publish_status()
+{
+  rcl_action_goal_status_array_t status_message =
+    rcl_action_get_zero_initialized_goal_status_array();
+  rcl_ret_t ret = rcl_action_get_goal_status_array(rcl_action_server_.get(), &status_message);
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed get goal status array");
+  }
+
+  ret = rcl_action_publish_status(rcl_action_server_.get(), &status_message);
+
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed publish goal status array");
+  }
+}
+
+void
+ActionServer::notify_goal_done()
+{
+  rcl_ret_t ret = rcl_action_notify_goal_done(rcl_action_server_.get());
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to notfiy action server of goal done");
+  }
+}
+
+bool
+ActionServer::goal_exists(py::object pygoal_info)
+{
+  auto goal_info = convert_from_py(pygoal_info);
+  rcl_action_goal_info_t * goal_info_type = static_cast<rcl_action_goal_info_t *>(goal_info.get());
+  if (!goal_info_type) {
```

Comment:
@ahcorde `goal_info` cannot ever be `nullptr`, [`convert_from_py` would've thrown](https://github.com/ros2/rclpy/blob/1ab8af41aaa8b085d011b0a695b7ee7ffe800f68/rclpy/src/rclpy/utils.cpp#L78).